### DISCUSSION TITLE
How can I cut down on duplicated native library versions?

### DISCUSSION THREAD

**Discussion by netherfoam** [pyinstaller/pyinstaller#7778] [Help] [answered]
I've generated an app.spec file, packaged as --onedir, and it works nicely, except for a few big, duplicated native libraries that are similarly versioned. For example, I have included rasterio, netcdf4 and pyproj in my project:

requirements.txt
```
...
rasterio==1.3.8 
netcdf4==1.6.4 
pyproj==3.6.0
...
```

but that leaves me with 3 libcurl native libraries. From the analysis (`a.binaries`):

```python
('libcurl-100461b2.so.4.8.0', '/root/.pyenv/versions/3.10.11/lib/python3.10/site-packages/pyproj/../pyproj.libs/libcurl-100461b2.so.4.8.0', 'BINARY'),
('libcurl-47074032.so.4.8.0', '/root/.pyenv/versions/3.10.11/lib/python3.10/site-packages/netCDF4/../netCDF4.libs/libcurl-47074032.so.4.8.0', 'BINARY'),
('libcurl-b6db4815.so.4.7.0', '/root/.pyenv/versions/3.10.11/lib/python3.10/site-packages/rasterio/../rasterio.libs/libcurl-b6db4815.so.4.7.0','BINARY'),
```

And naturally, the output:
```
git:(master) ls dist/linux/gridnix/libcurl*
dist/linux/gridnix/libcurl-100461b2.so.4.8.0  dist/linux/gridnix/libcurl-47074032.so.4.8.0  dist/linux/gridnix/libcurl-b6db4815.so.4.7.0
```

These are all very similar versions, and replacing one version with a symlink to another worked at first glance. I find someone else has had a similar issue, but instead through Conda: https://github.com/orgs/pyinstaller/discussions/6389 

* How can I reduce the number of similar-versioned native libraries I am bundling with PyInstaller?
* Or, is PyInstaller the right place to tackle this problem? Should I be modifying / compiling the dependencies I am currently installing from pip?

I've little experience with the Python and native library ecosystems, any broad advice on what direction to approach this problem from is appreciated!

**Comment by bwoodsend** [comment, marked as answer]
You're playing with fire here. There's no built in way of doing this in PyInstaller because it's not something we'd want to encourage ABI naive users into doing.

PyPI has a fairly strict policy on statically linking dependencies which is at odds with what you want so pip and PyPI are not your friends here. You might have better luck using a native Linux package manager (Fedora are quite aggressive about dynamic linking – they'd be my first stop) then run PyInstaller in the system Python environment.

All in all though, I think this is more trouble than it's worth.

**Reply by netherfoam** [reply]
Perfect, I’ll have a brief look into the native package manager.  If not, it is what it is.

Thanks a lot for the advice, you have likely saved me days of hair pulling!
